File: senclaw-extension-chrome/src/agent/CrawlEngine.ts
// BFS crawl engine: starts from a URL, follows links matching patterns.
import type { CrawlConfig, CrawlPageResult, JobId } from '../types/protocol';

interface CrawlJobState {
  config: CrawlConfig;
  visited: Set<string>;
  queue: string[];
  results: CrawlPageResult[];
  pagesCrawled: number;
  status: 'running' | 'paused' | 'completed' | 'stopped';
  startTime: number;
}

type ProgressCallback = (jobId: JobId, pagesCrawled: number, pagesTotal: number, currentUrl: string) => void;
type ResultCallback = (jobId: JobId, pageResult: CrawlPageResult) => void;
type CompleteCallback = (jobId: JobId, totalPages: number, durationMs: number) => void;

export class CrawlEngine {
  private jobs: Map<JobId, CrawlJobState> = new Map();
  private onProgress: ProgressCallback | null = null;
  private onResult: ResultCallback | null = null;
  private onComplete: CompleteCallback | null = null;

  setProgressCallback(cb: ProgressCallback): void { this.onProgress = cb; }
  setResultCallback(cb: ResultCallback): void { this.onResult = cb; }
  setCompleteCallback(cb: CompleteCallback): void { this.onComplete = cb; }

  async start(config: CrawlConfig): Promise<void> {
    const job: CrawlJobState = {
      config,
      visited: new Set(),
      queue: [config.start_url],
      results: [],
      pagesCrawled: 0,
      status: 'running',
      startTime: Date.now(),
    };

    this.jobs.set(config.job_id, job);
    await this.runLoop(config.job_id);
  }

  pause(jobId: JobId): void {
    const job = this.jobs.get(jobId);
    if (job) job.status = 'paused';
  }

  resume(jobId: JobId): void {
    const job = this.jobs.get(jobId);
    if (job) {
      job.status = 'running';
      this.runLoop(jobId);
    }
  }

  stop(jobId: JobId): void {
    const job = this.jobs.get(jobId);
    if (job) job.status = 'stopped';
  }

  getStatus(jobId: JobId): { status: string; pagesCrawled: number; results: CrawlPageResult[] } | null {
    const job = this.jobs.get(jobId);
    if (!job) return null;
    return {
      status: job.status,
      pagesCrawled: job.pagesCrawled,
      results: job.results,
    };
  }

  private async runLoop(jobId: JobId): Promise<void> {
    const job = this.jobs.get(jobId);
    if (!job) return;

    while (job.queue.length > 0 && job.status === 'running') {
      if (job.pagesCrawled >= job.config.max_pages) {
        job.status = 'completed';
        break;
      }

      const url = job.queue.shift()!;
      if (job.visited.has(url)) continue;

      // Same-domain check
      if (job.config.same_domain) {
        try {
          const startHost = new URL(job.config.start_url).host;
          const currentHost = new URL(url).host;
          if (startHost !== currentHost) continue;
        } catch { continue; }
      }

      // Exclude pattern check
      if (job.config.exclude_patterns.some(p => {
        try { return new RegExp(p).test(url); } catch { return url.includes(p); }
      })) continue;

      // Link pattern check (if specified)
      if (job.config.link_patterns.length > 0 &&
        !job.config.link_patterns.some(p => {
          try { return new RegExp(p).test(url); } catch { return url.includes(p); }
        })) continue;

      try {
        const result = await this.crawlPage(job, url);
        job.visited.add(url);
        job.results.push(result);
        job.pagesCrawled++;

        this.onResult?.(jobId, result);
        this.onProgress?.(
          jobId,
          job.pagesCrawled,
          Math.min(job.config.max_pages, job.queue.length + job.pagesCrawled),
          url,
        );

        // Discover new links if not at max depth
        if (result.depth < job.config.depth) {
          for (const link of result.links_found ? [] : []) {
            if (!job.visited.has(link) && !job.queue.includes(link)) {
              job.queue.push(link);
            }
          }
        }
      } catch (e) {
        console.error(`[CrawlEngine] Error crawling ${url}:`, e);
      }

      // Polite delay
      await this.sleep(job.config.wait_between_pages_ms || 1000);
    }

    if (job.status === 'running') {
      job.status = 'completed';
    }

    this.onComplete?.(jobId, job.pagesCrawled, Date.now() - job.startTime);
    this.jobs.delete(jobId);
  }

  private async crawlPage(job: CrawlJobState, url: string): Promise<CrawlPageResult> {
    // Create a background tab for crawling
    const tab = await chrome.tabs.create({ url, active: false });
    const tabId = tab.id!;

    // Wait for page load
    await new Promise<void>((resolve) => {
      const listener = (_tabId: number, changeInfo: chrome.tabs.TabChangeInfo) => {
        if (_tabId === tabId && changeInfo.status === 'complete') {
          chrome.tabs.onUpdated.removeListener(listener);
          resolve();
        }
      };
      chrome.tabs.onUpdated.addListener(listener);
      // Timeout
      setTimeout(() => {
        chrome.tabs.onUpdated.removeListener(listener);
        resolve();
      }, job.config.per_page_timeout_ms || 10000);
    });

    // Extract content
    let textContent = '';
    let linksFound = 0;
    try {
      const result = await chrome.tabs.sendMessage(tabId, { type: 'ExtractText' });
      textContent = result?.text ?? '';
    } catch {
      // Content script may not be loaded
    }

    try {
      const linkResult = await chrome.tabs.sendMessage(tabId, { type: 'ExtractLinks' });
      linksFound = linkResult?.links?.length ?? 0;
    } catch {
      // Ignore
    }

    // Close the crawl tab
    await chrome.tabs.remove(tabId);

    return {
      url,
      title: '',
      text_content: textContent.slice(0, 10000),
      links_found: linksFound,
      depth: job.visited.size,
      crawled_at: new Date().toISOString(),
    };
  }

  private sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}

File: src/browser/crawl_engine.rs
//! Crawl engine — BFS-based deep crawl scheduler.
//!
//! Manages crawl jobs: starting from a URL, following links matching patterns,
//! respecting depth limits, rate limiting between requests.

use std::collections::{HashMap, HashSet};
use std::sync::Arc;
use tokio::sync::RwLock;

use super::types::*;

/// A crawl job's internal state.
#[derive(Debug, Clone)]
struct CrawlJobState {
    config: CrawlConfig,
    visited: HashSet<String>,
    queue: Vec<String>,
    results: Vec<CrawlPageResult>,
    pages_crawled: u16,
    status: JobStatus,
    created_at: std::time::Instant,
}

#[derive(Debug, Clone, PartialEq)]
enum JobStatus {
    Running,
    Paused,
    Completed,
    Stopped,
}

/// Thread-safe crawl engine.
#[derive(Clone)]
pub struct CrawlEngine {
    jobs: Arc<RwLock<HashMap<JobId, CrawlJobState>>>,
}

impl CrawlEngine {
    pub fn new() -> Self {
        Self {
            jobs: Arc::new(RwLock::new(HashMap::new())),
        }
    }

    /// Register a new crawl job.
    pub async fn create_job(&self, config: CrawlConfig) -> JobId {
        let job_id = config.job_id.clone();
        let mut jobs = self.jobs.write().await;
        jobs.insert(
            job_id.clone(),
            CrawlJobState {
                visited: HashSet::new(),
                queue: vec![config.start_url.clone()],
                results: Vec::new(),
                pages_crawled: 0,
                status: JobStatus::Running,
                created_at: std::time::Instant::now(),
                config,
            },
        );
        job_id
    }

    /// Get the next URL to crawl for a job (BFS order).
    pub async fn next_url(&self, job_id: &str) -> Option<String> {
        let mut jobs = self.jobs.write().await;
        let job = jobs.get_mut(job_id)?;

        while let Some(url) = job.queue.pop() {
            if job.visited.contains(&url) {
                continue;
            }
            if job.pages_crawled >= job.config.max_pages {
                job.status = JobStatus::Completed;
                return None;
            }

            // Same-domain check
            if job.config.same_domain {
                if let (Some(start_host), Some(current_host)) =
                    (url_host(&job.config.start_url), url_host(&url))
                {
                    if start_host != current_host {
                        continue;
                    }
                }
            }

            // Exclude pattern check
            if job
                .config
                .exclude_patterns
                .iter()
                .any(|p| regex_match(p, &url))
            {
                continue;
            }

            // Link pattern check (if patterns specified, URL must match at least one)
            if !job.config.link_patterns.is_empty()
                && !job
                    .config
                    .link_patterns
                    .iter()
                    .any(|p| regex_match(p, &url))
            {
                continue;
            }

            job.visited.insert(url.clone());
            return Some(url);
        }

        None
    }

    /// Add a crawled page result to the job.
    pub async fn add_result(&self, job_id: &str, result: CrawlPageResult) {
        let mut jobs = self.jobs.write().await;
        if let Some(job) = jobs.get_mut(job_id) {
            job.results.push(result);
            job.pages_crawled += 1;
        }
    }

    /// Enqueue new URLs discovered on a page.
    pub async fn enqueue_urls(&self, job_id: &str, urls: Vec<String>) {
        let mut jobs = self.jobs.write().await;
        if let Some(job) = jobs.get_mut(job_id) {
            for url in urls {
                if !job.visited.contains(&url) && !job.queue.contains(&url) {
                    job.queue.push(url);
                }
            }
        }
    }

    /// Update crawl progress (called from extension progress events).
    pub async fn update_progress(
        &self,
        job_id: &str,
        pages_crawled: u16,
        pages_total: u16,
        _current_url: &str,
    ) {
        let jobs = self.jobs.read().await;
        if let Some(job) = jobs.get(job_id) {
            tracing::debug!("[CrawlEngine] Job {job_id}: {pages_crawled}/{pages_total} pages");
        }
    }

    /// Mark a crawl job as completed.
    pub async fn mark_complete(&self, job_id: &str) {
        let mut jobs = self.jobs.write().await;
        if let Some(job) = jobs.get_mut(job_id) {
            job.status = JobStatus::Completed;
        }
    }

    /// Pause a crawl job.
    pub async fn pause(&self, job_id: &str) -> bool {
        let mut jobs = self.jobs.write().await;
        if let Some(job) = jobs.get_mut(job_id) {
            job.status = JobStatus::Paused;
            true
        } else {
            false
        }
    }

    /// Resume a crawl job.
    pub async fn resume(&self, job_id: &str) -> bool {
        let mut jobs = self.jobs.write().await;
        if let Some(job) = jobs.get_mut(job_id) {
            job.status = JobStatus::Running;
            true
        } else {
            false
        }
    }

    /// Stop a crawl job.
    pub async fn stop(&self, job_id: &str) -> bool {
        let mut jobs = self.jobs.write().await;
        if let Some(job) = jobs.get_mut(job_id) {
            job.status = JobStatus::Stopped;
            true
        } else {
            false
        }
    }

    /// Get the status of a crawl job.
    pub async fn get_status(&self, job_id: &str) -> Option<CrawlJobStatus> {
        let jobs = self.jobs.read().await;
        let job = jobs.get(job_id)?;
        let status_str = match job.status {
            JobStatus::Running => "running",
            JobStatus::Paused => "paused",
            JobStatus::Completed => "completed",
            JobStatus::Stopped => "stopped",
        };
        Some(CrawlJobStatus {
            job_id: job_id.to_owned(),
            status: status_str.to_owned(),
            pages_crawled: job.pages_crawled,
            pages_total: job.config.max_pages,
            results: job.results.clone(),
        })
    }

    /// List all active crawl job IDs.
    pub async fn list_jobs(&self) -> Vec<JobId> {
        self.jobs.read().await.keys().cloned().collect()
    }
}

impl Default for CrawlEngine {
    fn default() -> Self {
        Self::new()
    }
}

/// Extract the host from a URL.
fn url_host(url: &str) -> Option<String> {
    let without_proto = url
        .strip_prefix("https://")
        .or_else(|| url.strip_prefix("http://"))
        .unwrap_or(url);
    without_proto.split('/').next().map(|s| s.to_lowercase())
}

/// Simple regex match (supports basic glob-style patterns and regex).
fn regex_match(pattern: &str, text: &str) -> bool {
    // Try regex first
    if let Ok(re) = regex::Regex::new(pattern) {
        return re.is_match(text);
    }
    // Fall back to substring match
    text.contains(pattern)
}
